aci/file.go

```diff
@@ -100,3 +107,87 @@ func XzReader(r io.Reader) io.ReadCloser {
 
 	return rpipe
 }
+
+// ManifestFromImage creates a new schema.ImageManifest from the given ACI image.
+func ManifestFromImage(rs io.ReadSeeker) (*schema.ImageManifest, error) {
+	var im schema.ImageManifest
+
+	tr, err := NewCompressedTarReader(rs)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			// end of tar archive
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if hdr.Name == "manifest" {
+			data, err := ioutil.ReadAll(tr)
+			if err != nil {
+				return nil, err
+			}
+			if err := im.UnmarshalJSON(data); err != nil {
+				return nil, err
+			}
+			break
+		}
+	}
+	return &im, nil
+}
+
+// NewCompressedTarReader creates a new tar.Reader reading from the given ACI image.
+func NewCompressedTarReader(rs io.ReadSeeker) (*tar.Reader, error) {
+	cr, err := NewCompressedReader(rs)
+	if err != nil {
+		return nil, err
+	}
+	return tar.NewReader(cr), nil
+}
+
+// NewCompressedReader creates a new io.Reader from the given ACI image.
+func NewCompressedReader(rs io.ReadSeeker) (io.Reader, error) {
```

Comment:
Why do you need to rewind at the start of this function? Also, since this function returns a Reader, the buffered approach is fine IMHO (so you don't need to rewind after sniffing the type).